logging/structured/src/main/scala/tofu/logging/LoggingMid.scala

```diff
@@ -0,0 +1,157 @@
+package tofu.logging
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+import tofu.syntax.monadic._
+import tofu.higherKind.Mid
+import cats.Monad
+import tofu.higherKind.derived.HigherKindedMacros
+import tofu.Errors
+import tofu.syntax.handle._
+
+/** Logging middleware
+  * Alg[LoggingMid] is a special form of implicit evidence of injectable logging support
+  * generally you don't need `Logging` instance to derive this
+  * so choice of logging postponed until this middleware is attached to the core instance
+  */
+abstract class LoggingMid[A] {
+  def around[F[_]: Monad: LoggingBase](fa: F[A]): F[A]
+
+  def toMid[F[_]: Monad: LoggingBase]: Mid[F, A] = around(_)
+}
+
+object LoggingMid extends LoggingMidBuilder.DefaultImpl {
+  type Result[A] = LoggingMid[A]
+  def instance[U[_[_]]]: U[LoggingMid] = macro HigherKindedMacros.factorizeThis[U]
+
+  type Of[U[_[_]]] = U[LoggingMid]
+}
+
+/** Logging middleware generator */
+trait LoggingMidBuilder {
```

Comment:
Somebody once told me that mixing up "utility" code like all those bulders and mutable stuff together with actual API  hurts readability and discoverability. 
One can almost sink in all this stuff.
Is it possible to hide it in some folders or in other files to keep things clean?

WDYT?

Reply:
This is not a utility code.
Currently I have a habit of hiding utility things in the `impl` package.
But this one is something like a configuration as a code interface, one can easily craft its own logging mid derivation using this builder.